File: squareMult.cc
#include <iostream>
#include <vector>
#include <unordered_set>
using namespace std;

vector<bool>primes;
int n;
vector<int>b;
vector<int>a;
vector<vector<int>>v;
vector<bool>used;

vector<bool>provPrime;
void getPrimes(int x) {
	provPrime = vector<bool>(x, true);
	for (int i = 2; i < x; ++i) {
		if (provPrime[i]) {
			if (i >= x/2) primes[i] = true;
			for (int j = i*2; j < x; j += i) provPrime[j] = false;
		}
	}
}

int maxim = -10;
int provMul = -1;
int sols = 0;
void sol(int x, int y, bool f) {
	if (sols > maxim+10) {
		maxim = sols;
		cout << maxim << endl;
	}
	if (x == n && y == n) {
		++sols;
		return;
	}
	if (y == n) {provMul = a[x]; return sol(x+1, x, !f);}
	if (x == n) {
		if (a[y] != b[y]) return;
		return sol(y+1, y+1, !f);
	}
	if (x == y) {
		v[x][y] = 1;
		if (f) {
			return sol(x, y+1, f);
		}
		else {
			return sol(x+1, y, f);
		}
	}
	if (f) {
		for (int i = 1; i <= n*n; ++i) {
			if (primes[i]) continue;
			if (!used[i]) {
				used[i] = true;
				v[x][y] = i;
				a[x] *= i;
				b[y] *= i;
				sol(x, y+1, f);
				a[x] /= i;
				b[y] /= i;
				used[i] = false;
			}
		}
	}
	else {
		for (int i = 1; i <= n*n; ++i) {
			if (primes[i]) continue;
			if (provMul%i != 0) continue;
			if (!used[i]) {
				provMul /= i;
				used[i] = true;
				v[x][y] = i;
				a[x] *= i;
				b[y] *= i;
				sol(x+1, y, f);
				a[x] /= i;
				b[y] /= i;
				used[i] = false;
				provMul *= i;
			}
		}
	}
}


int main(){
	ios_base::sync_with_stdio(0);
	cin.tie(0);
	cout.tie(0);
	cin >> n;
	a = vector<int>(n, 1);
	b = vector<int>(n, 1);
	used = vector<bool>(n*n+1, 0);
	primes = vector<bool>(n*n+1, 0);
	getPrimes(n*n);
	v = vector<vector<int>>(n, vector<int>(n, -1));
	sol(0,0,true);
	cout << sols << endl;
}

File: squareMult.py
def isPrime(x):
	if x<2:
		return False
	i = 2
	while i*i<=x:
		if x%i == 0:
			return False
		i += 1
	return True

def getPrimes(n):
	primes = []
	for i in range(n//2+1, n):
		if isPrime(i):
			primes.append(i)
	return primes

n=4
a = [1 for x in range(n)]
b = [1 for x in range(n)]
v = [[-1 for x in range(n)] for y in range(n)]
s = {n*n+1}

def check(x):
	if a[x]==b[x]:
		return True
	return False


def sol(x,y,f):
	if x==n and y==n:
		return True
	if y==n:
		return sol(x+1, x, not f)
	if x==n:
		if not check(y):
			return False
		return sol(y+1, y+1, not f)
	if x==y:
		v[x][y] = 1
		if f:
			return sol(x, y+1, f)
		else:
			return sol(x+1, y, f)
	for i in range(n*n+1):
		if i == 0:
			continue
		if i in primes:
			continue
		if not i in s:
			s.add(i)
			v[x][y] = i
			a[x] *= i
			b[y] *= i
			if f:
				if sol(x, y+1, f):
					return True
			else:
				if sol(x+1, y, f):
					return True
			v[x][y] = -1
			a[x] = a[x] // i
			b[y] = b[y] // i
			s.discard(i)
	return False



primes = getPrimes(n*n)
if sol(0, 0, True):
	for i in v:
		print(i)
else:
	print("no sol")